session-replay/Sources/DatadogSessionReplay/Processor/Privacy/TextObfuscator.swift

```diff
@@ -0,0 +1,32 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+internal struct TextObfuscator {
+    /// The character to mask text with.
+    let maskCharacter: UnicodeScalar = "x"
+
+    /// Masks given `text` by replacing all not whitespace characters with `"x"`.
+    /// - Parameter text: the text to be masked
+    /// - Returns: masked text
+    func mask(text: String) -> String {
+        var masked = ""
+
+        var iterator = text.unicodeScalars.makeIterator()
+
+        while let nextScalar = iterator.next() {
+            switch nextScalar {
+            case " ", "\n", "\r", "\t":
+                masked.unicodeScalars.append(nextScalar)
+            default:
+                masked.unicodeScalars.append(maskCharacter)
+            }
+        }
+
+        return masked
+    }
```

Comment:
I see 2 things that could be improved in this method:

1. The [Unicode.Scalar.Properties/isWhitespace](https://developer.apple.com/documentation/swift/unicode/scalar/properties-swift.struct/iswhitespace) can tell you if the character should be treated as whitespace.

2. The `String.UnicodeScalarView.Iterator` is required by the `Sequence` protocol that enable the `for in` statement. It also comes with a bunch of extensions such as `map`. It's all based on the iterator, so the perf should be the same.

This could become something like:
```swift
func mask(text: String) -> String {
    let unicodeScalars = text.unicodeScalars.map { scalar in
        scalar.properties.isWhitespace ? scalar : maskCharacter
    }

    let unicodeScalarView = String.UnicodeScalarView(unicodeScalars)
    return String(unicodeScalarView)
}
```

Reply:
It's interesting feedback, so I gave it a try with some benchmarks 🚤. Unfortunately the time performance is ~4x worse than current, iterator-based implementation. For that reason, I'd rather stick to existing implementation, even if `isWhitespace` looks more portable (rationale: `" ", "\n", "\r", "\t"` should cover majority of use cases and if not, it scales up pretty easy).

**More explanation:**

I tried 2 variants with `mask()` (original) and `mask2()` (new) implementations:
- big: 40K characters-long text, 1K measures
- small: 40 characters-long text, 1K measures

Getting:
- big:
   - `mask()` - 2.28 ms
   - `mask2()` - 10.04 ms
- small:
   - `mask()` - 0.0047 ms
   - `mask2()` - 0.0145 ms

When tracking it down in profiler, I found that both new elements contribute to this slower performance, so neither using `.map {}` nor `.isWhitespace` will improve speed.
* `mask()` trace:
<img width="736" alt="Screenshot 2022-10-19 at 11 07 00" src="https://user-images.githubusercontent.com/2358722/196655655-ef97195e-a11d-4d75-b054-a4b25fd5ec34.png">

* `mask2()` trace:
<img width="998" alt="Screenshot 2022-10-19 at 11 07 29" src="https://user-images.githubusercontent.com/2358722/196655739-7b5b13b6-e2f7-4e38-9302-fc6d549747bb.png">

I didn't spent much time in digging more, but at the high level:
* `.map {}` seems to be adding quite a few routines related to generic sequences
* `.isWhitespace` leads to [`_swift_stdlib_getBinaryProperties`](https://github.com/apple/swift/blob/5265ae757e9a8fce8b48a4c8e2b37dce7c0ddbcc/stdlib/public/core/UnicodeScalarProperties.swift#L125) which is quite expensive but we lose track of it (hard to say what's inside) - in contrast, the `case "\n"` is O(1) because scalars are `ExpressibleByUnicodeScalarLiteral` so it's done with basic memory arithmetics

Let me know what you think @maxep 🙂

Benchmark code for reference:
```swift
func testBenchmark1() {
    let text: String = .mockRandom(among: .alphanumericsAndWhitespace, length: 40_000)

    benchmark("mask()") {
        _ = obfuscator.mask(text: text)
    }
}

func testBenchmark2() {
    let text: String = .mockRandom(among: .alphanumericsAndWhitespace, length: 40_000)

    benchmark("mask2()") {
        _ = obfuscator.mask2(text: text)
    }
}

func benchmark(_ label: String, block: () -> Void) {
    var measures: [TimeInterval] = []

    for i in (0..<1_010) {
        let start = Date()
        block()

        if i >= 10 { // only measure after it is warmed-up
            let stop = Date()
            measures.append(stop.timeIntervalSince(start))
        }
    }

    let sum = measures.reduce(0, +)
    let avg = sum / Double(measures.count)

    print("⭐️ AVG in \(label): \(avg * 1_000) ms")
}
```

Reply:
Oh nice!! thanks for doing the benchmark 👍 

I'm surprised how heavy is the `Unicode.Scalar.Properties/isWhitespace` 😳
Looking at the profiles, it seems that the `masked.unicodeScalars.append(...)` is also expensive compared to the `map` (its few extra routines doesn't seem heavy).

Would you be able to test a `mask3`?
```swift
func mask(text: String) -> String {
    let unicodeScalars = text.unicodeScalars.map { scalar in
        switch scalar {
            case " ", "\n", "\r", "\t":
                return scalar
            default:
                return maskCharacter
            }
    }

    let unicodeScalarView = String.UnicodeScalarView(unicodeScalars)
    return String(unicodeScalarView)
}
```

Reply:
I tried, it's in between and closer to "bad" performance:
* big:
   * mask() - 2.28 ms
   * mask2() - 10.04 ms
   * **mask3() - 8.38 ms**
* small:
   * mask() - 0.0047 ms
   * mask2() - 0.0145 ms
   * **mask3() - 0.0133 ms**

Difference is that iterator-based approach leads to `String.UnicodeScalarView.append(_:)` in the call stack, whereas map-based calls `String.UnicodeScalarView.append(contentsOf:)` that is visibly more complex:

- iterator-based profile (`String.UnicodeScalarView.append(_:)`):
<img width="1094" alt="Screenshot 2022-10-20 at 10 40 25" src="https://user-images.githubusercontent.com/2358722/196900300-33c50a03-c971-4410-acf5-c0b13c2f254f.png">


- map-based profile (`String.UnicodeScalarView.append(contentsOf:)`):
<img width="1095" alt="Screenshot 2022-10-20 at 10 40 15" src="https://user-images.githubusercontent.com/2358722/196900333-7ec73856-adbd-4451-ab87-d87597958b51.png">

Again, I didn't dive too deep, but even a glance at expanded call tree suggests it's coming from `.map {}` acting too generic and bringing some underlying allocation cost:

<img width="868" alt="Screenshot 2022-10-20 at 10 48 49" src="https://user-images.githubusercontent.com/2358722/196902333-9b3b1574-32bb-4ab0-a5ad-95a78bacbdf1.png">

I also tried using `.isWhitespace` + iterator in `mask4()` resulting with also "worse" performance:
* big:
   * `mask4()` - 4.52 ms
* small
   * `mask4()` - 0.0065ms

All that said, as we're discussing implementation detail of performance-critical piece (it will be called **every frame for all texts**) we should go with the most performant approach. Would be a different choice if it's about architecture-level API where convenience is key. WDYT @maxep?

Reply:
Totally, we are aiming for the most performant one. I was expecting the `isWhitespace` to be a pre-filled boolean (or at least is simple binary check) and `map` to perform the same as using the iterator. But that's clearly not the case!
Lets stick with your original implementation then 👍 And thanks for testing that!!